Crawler: Fehler 500 während Aktualisierung des Suchindex wird nicht angezeigt

Wenn der Crawler mit der Funktion ``Den Suchindex aktualisieren`` gestartet wird und während des Crawlvorgangs beim Zugriff auf eine Seite ein Fehler 500 auftritt, wird keine Fehlermeldung angezeigt, sondern nur die Erfolgsmeldung ``x URL(s) erfolgreich indexiert. 0 fehlgeschlagen.`` ausgegeben.

I just tried using  http://httpstat.us/500 and I then get `[ERROR] Checked 0 link(s) successfully. 1 were broken.` - that looks correct to me?
Can you tell me how I can reproduce this?

--- a/core-bundle/src/Crawl/Escargot/Subscriber/SearchIndexSubscriber.php
+++ b/core-bundle/src/Crawl/Escargot/Subscriber/SearchIndexSubscriber.php
@@ -19,18 +19,21 @@ use Psr\Log\LoggerAwareInterface;
 use Psr\Log\LoggerAwareTrait;
 use Psr\Log\LogLevel;
 use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\Exception\HttpExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
 use Symfony\Contracts\HttpClient\ResponseInterface;
 use Symfony\Contracts\Translation\TranslatorInterface;
 use Terminal42\Escargot\CrawlUri;
 use Terminal42\Escargot\EscargotAwareInterface;
 use Terminal42\Escargot\EscargotAwareTrait;
+use Terminal42\Escargot\Subscriber\ExceptionSubscriberInterface;
 use Terminal42\Escargot\Subscriber\HtmlCrawlerSubscriber;
 use Terminal42\Escargot\Subscriber\RobotsSubscriber;
 use Terminal42\Escargot\Subscriber\SubscriberInterface;
 use Terminal42\Escargot\Subscriber\Util;
 use Terminal42\Escargot\SubscriberLoggerTrait;
 
-class SearchIndexSubscriber implements EscargotSubscriberInterface, EscargotAwareInterface, LoggerAwareInterface
+class SearchIndexSubscriber implements EscargotSubscriberInterface, EscargotAwareInterface, ExceptionSubscriberInterface, LoggerAwareInterface
 {
     use EscargotAwareTrait;
     use LoggerAwareTrait;
@@ -204,5 +207,22 @@ class SearchIndexSubscriber implements EscargotSubscriberInterface, EscargotAwar
         $result->addInfo('stats', $stats);
 
         return $result;
+    }
+
+    public function onTransportException(CrawlUri $crawlUri, TransportExceptionInterface $exception, ResponseInterface $response): void
+    {
+        $this->logError($crawlUri, 'Could not request properly: '.$exception->getMessage());
+    }
+
+    public function onHttpException(CrawlUri $crawlUri, HttpExceptionInterface $exception, ResponseInterface $response, ChunkInterface $chunk): void
+    {
+        $this->logError($crawlUri, 'HTTP Status Code: '.$response->getStatusCode());
+    }
+
+    private function logError(CrawlUri $crawlUri, string $message): void
+    {
+        ++$this->stats['error'];
+
+        $this->logWithCrawlUri($crawlUri, LogLevel::ERROR, sprintf('Broken link! %s.', $message));
     }
 }

--- a/core-bundle/tests/Crawl/Escargot/Subscriber/SearchIndexSubscriberTest.php
+++ b/core-bundle/tests/Crawl/Escargot/Subscriber/SearchIndexSubscriberTest.php
@@ -21,7 +21,11 @@ use PHPUnit\Framework\MockObject\MockObject;
 use PHPUnit\Framework\TestCase;
 use Psr\Log\LoggerInterface;
 use Psr\Log\LogLevel;
+use Symfony\Component\HttpClient\Chunk\LastChunk;
+use Symfony\Component\HttpClient\Exception\ClientException;
+use Symfony\Component\HttpClient\Exception\TransportException;
 use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\Exception\HttpExceptionInterface;
 use Symfony\Contracts\HttpClient\ResponseInterface;
 use Symfony\Contracts\Translation\TranslatorInterface;
 use Terminal42\Escargot\BaseUriCollection;
@@ -300,6 +304,152 @@ class SearchIndexSubscriberTest extends TestCase
             new IndexerException('Major failure!'),
             LogLevel::DEBUG,
             'Forwarded to the search indexer. Did not index because of the following reason: Major failure!',
+            ['ok' => 0, 'warning' => 0, 'error' => 2],
+            ['ok' => 0, 'warning' => 0, 'error' => 1],
+        ];
+    }
+
+    /**
+     * @dataProvider onTransportExceptionProvider
+     */
+    public function testOnTransportException(TransportException $exception, ResponseInterface $response, string $expectedLogLevel, string $expectedLogMessage, array $expectedStats, array $previousStats = []): void
+    {
+        $logger = $this->createMock(LoggerInterface::class);
+
+        if ('' !== $expectedLogLevel) {
+            $logger
+                ->expects($this->once())
+                ->method('log')
+                ->with(
+                    $expectedLogLevel,
+                    $expectedLogMessage,
+                    $this->callback(
+                        function (array $context) {
+                            $this->assertInstanceOf(CrawlUri::class, $context['crawlUri']);
+                            $this->assertSame(SearchIndexSubscriber::class, $context['source']);
+
+                            return true;
+                        }
+                    )
+                )
+            ;
+        } else {
+            $logger
+                ->expects($this->never())
+                ->method('log')
+            ;
+        }
+
+        $indexer = $this->createMock(IndexerInterface::class);
+        $indexer
+            ->expects($this->never())
+            ->method('index')
+        ;
+
+        $escargot = Escargot::create(new BaseUriCollection([new Uri('https://contao.org')]), new InMemoryQueue());
+        $escargot = $escargot->withLogger($logger);
+
+        $subscriber = new SearchIndexSubscriber($indexer, $this->getTranslator());
+        $subscriber->setEscargot($escargot);
+        $subscriber->setLogger(new SubscriberLogger($logger, \get_class($subscriber)));
+        $subscriber->onTransportException(new CrawlUri(new Uri('https://contao.org'), 0), $exception, $response);
+
+        $previousResult = null;
+
+        if (0 !== \count($previousStats)) {
+            $previousResult = new SubscriberResult(true, 'foobar');
+            $previousResult->addInfo('stats', $previousStats);
+        }
+
+        $result = $subscriber->getResult($previousResult);
+        $this->assertSame($expectedStats, $result->getInfo('stats'));
+    }
+
+    public function onTransportExceptionProvider(): \Generator
+    {
+        yield 'Test reports transport exception responses' => [
+            new TransportException('Could not resolve host or timeout'),
+            $this->getResponse(true, 404),
+            LogLevel::ERROR,
+            'Broken link! Could not request properly: Could not resolve host or timeout.',
+            ['ok' => 0, 'warning' => 0, 'error' => 2],
+            ['ok' => 0, 'warning' => 0, 'error' => 1],
+        ];
+    }
+
+    /**
+     * @dataProvider onHttpExceptionProvider
+     */
+    public function testOnHttpException(HttpExceptionInterface $exception, ResponseInterface $response, ChunkInterface $chunk, string $expectedLogLevel, string $expectedLogMessage, array $expectedStats, array $previousStats = []): void
+    {
+        $logger = $this->createMock(LoggerInterface::class);
+
+        if ('' !== $expectedLogLevel) {
+            $logger
+                ->expects($this->once())
+                ->method('log')
+                ->with(
+                    $expectedLogLevel,
+                    $expectedLogMessage,
+                    $this->callback(
+                        function (array $context) {
+                            $this->assertInstanceOf(CrawlUri::class, $context['crawlUri']);
+                            $this->assertSame(SearchIndexSubscriber::class, $context['source']);
+
+                            return true;
+                        }
+                    )
+                )
+            ;
+        } else {
+            $logger
+                ->expects($this->never())
+                ->method('log')
+            ;
+        }
+
+        $indexer = $this->createMock(IndexerInterface::class);
+        $indexer
+            ->expects($this->never())
+            ->method('index')
+        ;
+
+        $escargot = Escargot::create(new BaseUriCollection([new Uri('https://contao.org')]), new InMemoryQueue());
+        $escargot = $escargot->withLogger($logger);
+
+        $subscriber = new SearchIndexSubscriber($indexer, $this->getTranslator());
+        $subscriber->setEscargot($escargot);
+        $subscriber->setLogger(new SubscriberLogger($logger, \get_class($subscriber)));
+        $subscriber->onHttpException(new CrawlUri(new Uri('https://contao.org'), 0), $exception, $response, $chunk);
+
+        $previousResult = null;
+
+        if (0 !== \count($previousStats)) {
+            $previousResult = new SubscriberResult(true, 'foobar');
+            $previousResult->addInfo('stats', $previousStats);
+        }
+
+        $result = $subscriber->getResult($previousResult);
+        $this->assertSame($expectedStats, $result->getInfo('stats'));
+    }
+
+    public function onHttpExceptionProvider(): \Generator
+    {
+        yield 'Test reports responses that were not successful' => [
+            new ClientException($this->getResponse(true, 404)),
+            $this->getResponse(true, 404),
+            new LastChunk(),
+            LogLevel::ERROR,
+            'Broken link! HTTP Status Code: 404.',
+            ['ok' => 0, 'warning' => 0, 'error' => 1],
+        ];
+
+        yield 'Test reports responses that were not successful (with previous result)' => [
+            new ClientException($this->getResponse(true, 404)),
+            $this->getResponse(true, 404),
+            new LastChunk(),
+            LogLevel::ERROR,
+            'Broken link! HTTP Status Code: 404.',
             ['ok' => 0, 'warning' => 0, 'error' => 2],
             ['ok' => 0, 'warning' => 0, 'error' => 1],
         ];

